sklearn/datasets/tests/test_base.py

```diff
@@ -391,3 +396,257 @@ def test_fetch_remote_raise_warnings_with_invalid_url(monkeypatch):
         for r in record:
             assert str(r.message) == f"Retry downloading from url: {url}"
         assert len(record) == 3
+
+
+def test_derive_folder_and_filename_from_url():
+    folder, filename = _derive_folder_and_filename_from_url(
+        "https://example.com/file.tar.gz"
+    )
+    assert folder == "example.com"
+    assert filename == "file.tar.gz"
+
+    folder, filename = _derive_folder_and_filename_from_url(
+        "https://example.com/نمونه نماینده.data"
+    )
+    assert folder == "example.com"
+    assert filename == "نمونه-نماینده.data"
+
+    folder, filename = _derive_folder_and_filename_from_url(
+        "https://example.com/path/to/file.tar.gz"
+    )
+    assert folder == "example.com/path_to"
+    assert filename == "file.tar.gz"
+
+    folder, filename = _derive_folder_and_filename_from_url("https://example.com/")
+    assert folder == "example.com"
+    assert filename == "downloaded_file"
+
+    folder, filename = _derive_folder_and_filename_from_url("https://example.com")
+    assert folder == "example.com"
+    assert filename == "downloaded_file"
+
+    folder, filename = _derive_folder_and_filename_from_url(
+        "https://example.com/path/@to/data.json?param=value"
+    )
+    assert folder == "example.com/path_to"
+    assert filename == "data.json"
+
+    folder, filename = _derive_folder_and_filename_from_url(
+        "https://example.com/path/@@to/data.json#anchor"
+    )
+    assert folder == "example.com/path_to"
+    assert filename == "data.json"
+
+    folder, filename = _derive_folder_and_filename_from_url(
+        "https://example.com//some_file.txt"
+    )
+    assert folder == "example.com"
+    assert filename == "some_file.txt"
+
+    folder, filename = _derive_folder_and_filename_from_url(
+        "https://example.com/!.'.,/some_file.txt"
+    )
+    assert folder == "example.com"
+    assert filename == "some_file.txt"
+
+    folder, filename = _derive_folder_and_filename_from_url(
+        "https://example.com/a/!.'.,/b/some_file.txt"
+    )
+    assert folder == "example.com/a_b"
+    assert filename == "some_file.txt"
+
+    folder, filename = _derive_folder_and_filename_from_url("https://example.com/!.'.,")
+    assert folder == "example.com"
+    assert filename == "downloaded_file"
+
+    with pytest.raises(ValueError, match="Invalid URL"):
+        _derive_folder_and_filename_from_url("https:/../")
```

Comment:
Note:

```py
In [4]: _derive_folder_and_filename_from_url("http://sample/../test.txt")
Out[4]: ('sample', 'test.txt')
```

which I think is fine, but not tested?

Reply:
I tested more complex cases where "." are interleaved with other filtered characters but you are right, let's test this more obvious case.